Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h

```diff
@@ -50,21 +50,35 @@ class OpenXRCoordinator final : public PlatformXRCoordinator {
     void startSession(WebPageProxy&, WeakPtr<PlatformXRCoordinatorSessionEventClient>&&, const WebCore::SecurityOriginData&, PlatformXR::SessionMode, const PlatformXR::Device::FeatureList&) override;
     void endSessionIfExists(WebPageProxy&) override;
 
-    void scheduleAnimationFrame(WebPageProxy&, std::optional<PlatformXR::RequestData>&&, PlatformXR::Device::RequestFrameCallback&& onFrameUpdateCallback) override { onFrameUpdateCallback({ }); }
+    void scheduleAnimationFrame(WebPageProxy&, std::optional<PlatformXR::RequestData>&&, PlatformXR::Device::RequestFrameCallback&& onFrameUpdateCallback) override;
+    void submitFrame(WebPageProxy&) override;
 
 private:
     void createInstance();
     void createSessionIfNeeded();
     void initializeDevice();
     void initializeSystem();
+    void initializeBlendModes();
     void initializeGraphicsBinding();
     void collectViewConfigurations();
     WebCore::IntSize recommendedResolution() const;
 
-    void handleSessionStateChange();
+    struct Idle {
+    };
+    struct Active {
+        WeakPtr<PlatformXRCoordinatorSessionEventClient> sessionEventClient;
+        WebCore::PageIdentifier pageIdentifier;
+        Box<RenderState> renderState;
+        RefPtr<Thread> renderThread;
+    };
+    using State = Variant<Idle, Active>;
+
+    void handleSessionStateChange(Box<RenderState>);
     void endSessionIfExists(std::optional<WebCore::PageIdentifier>);
-    void pollEvents(std::atomic<bool>& terminateRequested);
+    enum class PollResult : bool;
```

Comment:
Why don't you just define it here?

Reply:
Well it's an implementation detail, I though it'd be a good idea to "hide" it in the header, but I can do it here if you prefer.